src/aggregators/directcr.jl

```diff
@@ -0,0 +1,178 @@
+"""
+Composition-Rejection Direct Method (DirectCR), implementation combining
+features from the original article:
+*A constant-time kinetic Monte Carlo algorithm for simulation of large biochemical reaction networks*,
+by A. Slepoy, A.P. Thompson and S.J. Plimpton, J. Chem. Phys, 128, 205101 (2008).
+and
+*Efficient Formulations for Exact Stochastic Simulation of Chemical Systems*,
+by S. Mauch and M. Stalzer, ACM Trans. Comp. Biol. and Bioinf., 8, No. 1, 27-35 (2010).
+"""
+
+const MINJUMPRATE = 2.0^exponent(1e-12)
```

Comment:
I don't get why this needs to have a minimum? It needs a binning rate, but why not have the first bin go from 0 to binrate, then exponentiate from there?

Reply:
binrate is what I'm calling `minrate` (which gets set to `MINJUMPRATE` by default but can be passed into the aggregator as a kwarg). One should never sample from the bin with zero elements, but one could sample from the bin that has elements in (0,`minrate`). If there are lots of zero elements, and one is in a situation where the min non-zero bin is actually getting hit, then I was thinking that could make a performance difference. In practice I'm guessing for many "real-world" networks hitting the min non-zero bin will be exceedingly rare. (In general, one probably does want to pass in a maxpriority as the defaults could add way too many bins on the upper end, which is where you search down from.) 

I made the bin boundaries powers of two to make the bin calculation as simple as possible (i.e. so that the integer exponent gives the bin boundaries). Which now makes me realize I actually haven't checked set up the code to handle non-power of two min/max priorities passed in by the user... So that needs to get looked at before merging.

Reply:
I guess the argument if you're actually hitting the min non-zero bin a lot would be that you should've given a smaller `minrate` from the start. So I think I'll follow your suggestion and get rid of the (0,minrate) bin.

Reply:
Oh I was actually looking for it. I think you need the (0,minrate) bin. It wasn't clear to me from the code/paper it was in there. I think with enough reactions it would bias it to not be included.

Reply:
Right we're discussing slightly different issues I think.

I had both a zero only bin and then a (0,minrate) bin (open at the ends). One could alternatively have just one [0,minrate) bin. The latter would eliminate an if statement in the mapping function, which might be a win over separating out the zero rates. But maybe that's an optimization to worry about later.

Reply:
Yeah, we can take a look at it later. I see what's going on now. Yeah, I was expecting `[0,minrate)` and was slightly confused by the explicit zero handling.

Reply:
I'll add some of these optimization comments to the SSA book keeping thread so they don't get forgotten.

Reply:
Just to add, I tried to comment the table datastructure code pretty thoroughly and write it a bit more generically since it should be re-usable as a component in other SSAs or by other people (or at least by me for spatial SSAs).